change real to float in cor of a single collection

bkamins opened the pull request:
As in `cor` we get square root I think it is safe to assume that the result should be floating point.

An example of current surprising behavior:
```
julia> cor([im])
true
```

nalimilan reviewed (approved):
Thanks!

nalimilan commented:
Though could you add a test to check the new behavior?

bkamins commented:
A good point - I have added the tests. They highlight the cases that I think are critical. A check that in all cases we produce what we think should be produced is welcome (I think they are not always intuitive, but I made the PR to be non-breaking in these corner cases except changing the type of `1` produced).

nalimilan commented on test/runtests.jl:
```diff
@@ -516,6 +516,20 @@ end
     end
+    
+    @test cor(Int[]) === 1.0
+    @test cor([im]) ===1.0 + 0.0im
```
```suggestion
    @test cor([im]) === 1.0 + 0.0im
```

nalimilan commented:
I just noticed `corm` and `corzm` have similar definitions. I guess these should be changed too, right?

nalimilan commented:
@stevengj You introduced this definition at https://github.com/JuliaLang/julia/pull/13348, replacing `float(one(T) * one(T))` with `real(one(T))`. Yet `cor([1im, 2im], [1im, 2im])` returns `1.0 + 0.0im`, not `1.0`. I guess nobody really tested nor care about this use case. What do you think we should do? We could also use `float(real(one(T))` to minimize breakage; what matters is that `cor([1, 2])` returns `1.0`, not `1`.

bkamins commented:
> I guess these should be changed too, right?

What do you mean exactly? I have already changed `corm` and `corzm` in this PR (and added tests for this).

nalimilan commented:
> What do you mean exactly? I have already changed `corm` and `corzm` in this PR (and added tests for this).

Woops, forget it, I was just confused.

nalimilan commented:
@stevengj Any chance you could find the time to look at this?
